servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapSingle.java

```diff
@@ -261,30 +272,25 @@ private void enqueueAndDrain(Object item) {
 
         private void drainPending() {
             long drainCount = 0;
-            Throwable delayedCause = null;
             boolean tryAcquire = true;
             while (tryAcquire && tryAcquireLock(emittingUpdater, this)) {
                 try {
                     Object t;
                     while ((t = pending.poll()) != null) {
                         ++drainCount;
-                        try {
-                            sendToTarget(t);
-                        } catch (Throwable cause) {
-                            delayedCause = catchUnexpected(delayedCause, cause);
-                        }
+                        sendToTarget(t);
+                    }
+                    // Call doDrainPostProcessing inside the lock so if an exception is thrown we can propagate an
+                    // error terminal downstream and ignore future signals.
+                    if (drainCount != 0) {
+                        doDrainPostProcessing(drainCount);
                     }
-                } finally {
-                    tryAcquire = !releaseLock(emittingUpdater, this);
+                } catch (Throwable cause) {
+                    onErrorHoldingLock(cause);
+                    return; // Poison emittingUpdater. We prematurely terminated, other signals should be ignored.
                 }
-            }
-
-            if (delayedCause != null) {
-                throwException(delayedCause);
-            }
-
-            if (drainCount != 0) {
-                doDrainPostProcessing(drainCount);
+                // Release lock after we handle errors, because error handling needs to poison the lock.
+                tryAcquire = !releaseLock(emittingUpdater, this);
```

Comment:
1. Looks like this `releaseLock` invocation can also be impacted by an exception thrown from `onErrorHoldingLock`.
2. While you are here, please make `enqueueFailed` method below `static`.

Reply:
> Looks like this releaseLock invocation can also be impacted by an exception thrown from onErrorHoldingLock.

lets discuss above in https://github.com/apple/servicetalk/pull/1614#discussion_r649654072

> While you are here, please make enqueueFailed method below static.

done